### ISSUE TITLE
[Bug] MPO训练时无法使用 gradient checkpointing

### ISSUE THREAD

**Issue by zzzz123-0708** [OpenGVLab/InternVL#755]
### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [ ] 2. The bug has not been fixed in the latest version.
- [ ] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.

### Describe the bug

使用sft训练正常使用gradient checkpointing，但是使用MPO训练报错ValueError: InternVLChatModel does not support gradient checkpointing.

我修改了sh
set -x

GPUS=${GPUS:-8}
BATCH_SIZE=${BATCH_SIZE:-128}
PER_DEVICE_BATCH_SIZE=${PER_DEVICE_BATCH_SIZE:-4}
GRADIENT_ACC=$((BATCH_SIZE / PER_DEVICE_BATCH_SIZE / GPUS))

export PYTHONPATH="${PYTHONPATH}:$(pwd)"
export MASTER_PORT=34223
export TF_CPP_MIN_LOG_LEVEL=3
export LAUNCHER=pytorch

OUTPUT_DIR='work_dirs/internvl_chat_v2_0/internvl2_8b_internlm2_7b_dynamic_res_mpo_full/ep5_bs64_5e-6_cot_3_all/'

if [ ! -d "$OUTPUT_DIR" ]; then
  mkdir -p "$OUTPUT_DIR"
fi

# number of gpus: 64
# batch size per gpu: ~4
# gradient accumulation steps: 1
# total batch size: ~256
# epoch: 8
torchrun \
  --nnodes=1 \
  --node_rank=0 \
  --master_addr=127.0.0.1 \
  --nproc_per_node=${GPUS} \
  --master_port=${MASTER_PORT} \
  internvl/train/internvl_chat_dpo.py \
  --model_name_or_path "/data3/.cache/hub/InternVL2-8B/" \
  --conv_style "internlm2-chat" \
  --output_dir ${OUTPUT_DIR} \
  --meta_path "./shell/data/data_info_MPO.jsonl" \
  --overwrite_output_dir True \
  --force_image_size 448 \
  --down_sample_ratio 0.5 \
  --drop_path_rate 0.1 \
  --pad2square False \
  --freeze_llm False \
  --freeze_mlp False \
  --freeze_backbone False \
  --vision_select_layer -1 \
  --use_data_resampling False \
  --dataloader_num_workers 8 \
  --bf16 True \
  --num_train_epochs 5 \
  --per_device_train_batch_size ${PER_DEVICE_BATCH_SIZE} \
  --gradient_accumulation_steps ${GRADIENT_ACC} \
  --evaluation_strategy "no" \
  --save_strategy "no" \
  --save_steps 100 \
  --save_total_limit 100 \
  --learning_rate 5e-6 \
  --weight_decay 0.05 \
  --warmup_ratio 0.03 \
  --lr_scheduler_type "cosine" \
  --logging_steps 1 \
  --max_seq_length 6144 \
  --do_train True \
  --grad_checkpoint True \
  --group_by_length False \
  --dynamic_image_size False \
  --use_thumbnail True \
  --ps_version 'v2' \
  --deepspeed "zero_stage3_config.json" \
  --report_to "tensorboard" \
  --loss_type sigmoid,bco_pair \
  --sigmoid_loss_weight 0.8 \
  --bco_pair_loss_weight 0.2 \
  --rpo_alpha 1 \
  2>&1 | tee -a "${OUTPUT_DIR}/training_log.txt"

是我修改的sh有问题吗？期待回复！！

### Reproduction

CUDA_VISIBLE_DEVICES=0,1,2,3 GPUS=4 PER_DEVICE_BATCH_SIZE=4 BATCH_SIZE=64 bash shell/internvl2.0_mpo/preference_optimization/internvl2_8b_internlm2_7b_dynamic_res_mpo_full.sh

### Environment

```Shell
Package                       Version              Editable project location
----------------------------- -------------------- --------------------------------------------------------------
absl-py                       1.4.0
addict                        2.4.0
aenum                         3.1.15
affine                        2.4.0
aiofiles                      22.1.0
aiohappyeyeballs              2.3.4
aiohttp                       3.10.0
aiosignal                     1.3.1
aiosqlite                     0.18.0
albumentations                1.2.1
aliyun-python-sdk-core        2.14.0
aliyun-python-sdk-kms         2.16.2
altair                        5.1.2
altgraph                      0.17.4
annotated-types               0.6.0
antlr4-python3-runtime        4.9.3
anyio                         3.7.1
appdirs                       1.4.4
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
asttokens                     2.0.5
async-timeout                 4.0.3
attrs                         23.1.0
audioread                     3.0.1
auto_gptq                     0.7.1
av                            12.3.0
Babel                         2.11.0
backcall                      0.2.0
beautifulsoup4                4.12.2
black                         23.7.0
bleach                        4.1.0
blessed                       1.20.0
blinker                       1.7.0
blis                          0.7.11
brotlipy                      0.7.0
cachetools                    5.3.1
catalogue                     2.0.10
certifi                       2024.8.30
cffi                          1.15.1
chardet                       5.2.0
charset-normalizer            3.4.0
cityscapesScripts             2.2.2
click                         8.1.3
click-plugins                 1.1.1
cligj                         0.7.2
cloudpickle                   2.2.1
cmake                         3.30.1
cn-clip                       1.5.1
codecov                       2.1.13
colorama                      0.4.6
coloredlogs                   15.0.1
comm                          0.1.2
confection                    0.1.5
contourpy                     1.1.0
coverage                      7.2.7
crcmod                        1.7
crowdposetools                2.0
cryptography                  39.0.1
cssselect2                    0.7.0
cycler                        0.11.0
cymem                         2.0.8
Cython                        3.0.2
cytoolz                       0.12.0
dask                          2023.4.1
debugpy                       1.5.1
decorator                     5.1.1
defusedxml                    0.7.1
detectron2                    0.6
diffusers                     0.30.3
dill                          0.3.4
diskcache                     5.6.3
distro                        1.9.0
docker                        7.1.0
docker-pycreds                0.4.0
docstring_parser              0.16
easydict                      1.12
einops                        0.6.1
einops-exts                   0.0.4
emoji                         2.2.0
entrypoints                   0.4
et-xmlfile                    1.1.0
exceptiongroup                1.1.1
executing                     0.8.3
fairscale                     0.4.13
fastapi                       0.104.0
fastjsonschema                2.16.2
fasttext-wheel                0.9.2
ffmpy                         0.3.1
filelock                      3.16.1
fiona                         1.9.5
fire                          0.6.0
flake8                        6.0.0
FLAML                         2.1.2
flash-attn                    2.6.2
Flask                         3.0.2
Flask-Cors                    4.0.0
flatbuffers                   24.3.25
fonttools                     4.40.0
frozendict                    2.4.4
frozenlist                    1.4.1
fsspec                        2024.9.0
ftfy                          6.1.3
future                        0.18.3
fvcore                        0.1.5.post20221221
gdown                         5.2.0
gekko                         1.2.1
gensim                        4.3.3
geopandas                     0.13.2
gitdb                         4.0.11
GitPython                     3.1.43
google-auth                   2.19.1
google-auth-oauthlib          1.0.0
gpustat                       1.1
GPUtil                        1.4.0
gradio                        3.50.2
gradio_client                 0.6.1
graphsurgeon                  0.4.6
grpcio                        1.54.2
h11                           0.14.0
h5py                          3.9.0
hjson                         3.1.0
html5lib                      1.1
httpcore                      0.18.0
httpx                         0.25.0
huggingface-hub               0.26.1
humanfriendly                 10.0
hydra-core                    1.3.2
idna                          3.10
imagecodecs                   2023.1.23
imagededup                    0.3.2
imageio                       2.31.1
imagesize                     1.4.1
imgaug                        0.4.0
imgviz                        1.7.5
importlib_metadata            8.5.0
importlib-resources           5.12.0
iniconfig                     2.0.0
interrogate                   1.5.0
iopath                        0.1.9
ipykernel                     6.19.2
ipython                       8.12.0
ipython-genutils              0.2.0
ipywidgets                    8.0.4
itsdangerous                  2.1.2
jedi                          0.18.1
Jinja2                        3.1.2
jmespath                      0.10.0
joblib                        1.3.2
jpeg4py                       0.1.4
json5                         0.9.6
jsonargparse                  4.32.0
jsonlines                     4.0.0
jsonschema                    4.17.3
jupyter                       1.0.0
jupyter_client                8.1.0
jupyter-console               6.6.3
jupyter_core                  5.3.0
jupyter-events                0.6.3
jupyter_server                2.5.0
jupyter_server_fileid         0.9.0
jupyter_server_terminals      0.4.4
jupyter_server_ydoc           0.8.0
jupyter-ydoc                  0.2.4
jupyterlab                    3.6.3
jupyterlab-pygments           0.1.2
jupyterlab_server             2.22.0
jupyterlab-widgets            3.0.5
kenlm                         0.2.0
kiwisolver                    1.4.4
kornia                        0.7.0
langcodes                     3.4.0
language_data                 1.2.0
lazy_loader                   0.4
Levenshtein                   0.25.1
librosa                       0.10.2.post1
linformer                     0.2.1
linkify-it-py                 2.0.3
lit                           18.1.8
llvmlite                      0.41.1
lmdb                          1.3.0
locket                        1.0.0
loguru                        0.7.2
lxml                          4.9.1
lz4                           4.3.3
mahotas                       1.4.13
Mako                          1.3.5
map-boxes                     1.0.5
marisa-trie                   1.2.0
Markdown                      3.4.3
markdown-it-py                3.0.0
markdown2                     2.5.0
MarkupSafe                    2.1.2
matplotlib                    3.7.1
matplotlib-inline             0.1.6
mccabe                        0.7.0
mdit-py-plugins               0.3.3
mdurl                         0.1.2
mistune                       0.8.4
mkl-fft                       1.3.8
mkl-random                    1.2.4
mkl-service                   2.4.0
mmdeploy                      1.3.1
mmdeploy-runtime-gpu          1.3.1
mmengine                      0.7.4
mmsegmentation                1.0.0                /home/tiancai/anaconda3/envs/torch/lib/python3.8/site-packages
model-index                   0.1.11
modelscope                    1.17.0
mpmath                        1.3.0
msgpack                       1.0.8
multidict                     6.0.5
multiprocess                  0.70.12
MultiScaleDeformableAttention 1.0
multitasking                  0.0.11
murmurhash                    1.0.10
mypy-extensions               1.0.0
nbclassic                     0.5.5
nbclient                      0.5.13
nbconvert                     6.5.4
nbformat                      5.7.0
nest-asyncio                  1.5.6
networkx                      3.1
nibabel                       5.1.0
ninja                         1.11.1.1
nlpaug                        1.1.11
nltk                          3.8.2
notebook                      6.5.2
notebook_shim                 0.2.2
numba                         0.58.1
numpy                         1.24.4
nvidia-cuda-runtime-cu12      12.6.77
nvidia-cufft-cu11             10.9.0.58
nvidia-ml-py                  11.525.112
nvitop                        1.3.2
oauthlib                      3.2.2
omegaconf                     2.3.0
onnx                          1.16.0
onnxruntime                   1.17.3
onnxruntime-gpu               1.19.2
open-clip-torch               2.16.0
openai                        1.35.14
opencv-contrib-python         4.8.0.76
opencv-python                 4.8.0.76
opencv-python-headless        4.8.1.78
openmim                       0.3.7
openpyxl                      3.1.5
openxlab                      0.0.27
optimum                       1.21.4
ordered-set                   4.1.0
orjson                        3.9.10
oss2                          2.17.0
packaging                     24.1
paho-mqtt                     1.6.1
pandas                        2.0.3
pandas-datareader             0.10.0
pandocfilters                 1.5.0
panopticapi                   0.1
parso                         0.8.3
partd                         1.4.0
pathlib_abc                   0.1.1
pathspec                      0.11.2
pathy                         0.11.0
pdfminer.six                  20231228
pdfplumber                    0.11.2
peewee                        3.17.6
peft                          0.12.0
pexpect                       4.8.0
pickleshare                   0.7.5
pillow                        10.4.0
pip                           23.0.1
pkgutil_resolve_name          1.3.10
platformdirs                  3.5.3
plotly                        5.23.0
pluggy                        1.0.0
ply                           3.11
pooch                         1.4.0
portalocker                   2.7.0
preshed                       3.0.9
prettytable                   3.8.0
progressbar                   2.5
prometheus-client             0.14.1
prompt-toolkit                3.0.36
protobuf                      3.20.3
psutil                        5.9.5
ptyprocess                    0.7.0
pure-eval                     0.2.2
py                            1.11.0
py-cpuinfo                    9.0.0
py4j                          0.10.9.7
pyaml                         20.4.0
pyarrow-hotfix                0.6
pyasn1                        0.5.0
pyasn1-modules                0.3.0
pyautogen                     0.2.32
pybind11                      2.13.1
pycocotools                   2.0
pycodestyle                   2.10.0
pycparser                     2.21
pycryptodome                  3.19.0
pycuda                        2024.1.2
pydantic                      2.4.2
pydantic_core                 2.10.1
pydeck                        0.9.1
pydot                         1.4.2
pydub                         0.25.1
pyflakes                      3.0.1
pygame                        2.5.2
Pygments                      2.15.1
pyinstaller                   6.3.0
pyinstaller-hooks-contrib     2023.12
pyOpenSSL                     23.0.0
pyparsing                     3.0.9
pypdfium2                     4.30.0
pyproj                        3.5.0
PyQt5                         5.15.10
PyQt5-Qt5                     5.15.2
PyQt5-sip                     12.13.0
pyquaternion                  0.9.9
pyrsistent                    0.18.0
pyshp                         2.3.1
PySocks                       1.7.1
pyspark                       3.5.2
pytest                        7.3.2
python-dateutil               2.8.2
python-docx                   1.1.2
python-dotenv                 1.0.1
python-json-logger            2.0.7
python-Levenshtein            0.25.1
python-multipart              0.0.6
pytools                       2024.1.14
pytz                          2023.3
PyWavelets                    1.4.1
PyYAML                        6.0.2
pyzmq                         25.1.0
qtconsole                     5.4.2
QtPy                          2.2.0
qudida                        0.0.4
rapidfuzz                     3.9.5
rasterio                      1.3.9
regex                         2024.9.11
reportlab                     4.0.6
requests                      2.32.3
requests-oauthlib             1.3.1
rfc3339-validator             0.1.4
rfc3986-validator             0.1.1
rich                          13.4.2
rouge                         1.0.1
rsa                           4.9
sacremoses                    0.1.1
safetensors                   0.4.5
schedule                      1.2.1
scikit-image                  0.19.3
scikit-learn                  1.2.2
scipy                         1.10.1
seaborn                       0.13.2
selectolax                    0.3.21
semantic-version              2.10.0
Send2Trash                    1.8.0
sentencepiece                 0.1.99
sentry-sdk                    2.12.0
setproctitle                  1.3.3
setuptools                    68.0.0
shapely                       2.0.1
shortuuid                     1.0.13
simhash-pybind                0.0.3
simple-aesthetics-predictor   0.1.2
simplejson                    3.19.1
sip                           6.6.2
six                           1.16.0
sklearn                       0.0.post9
smart-open                    6.4.0
smmap                         5.0.1
sniffio                       1.2.0
snuggs                        1.4.7
soundfile                     0.12.1
soupsieve                     2.4
soxr                          0.3.7
spacy-legacy                  3.0.12
spacy-loggers                 1.0.5
srsly                         2.4.8
stack-data                    0.2.0
starlette                     0.27.0
streamlit                     1.37.0
submitit                      1.4.5
svglib                        1.5.1
svgwrite                      1.4.3
swin-window-process           0.0.0
sympy                         1.12
tabulate                      0.9.0
tenacity                      8.5.0
tensorboard                   2.13.0
tensorboard-data-server       0.7.0
tensorboardX                  2.6.2.2
tensorrt-cu12                 10.5.0
tensorrt-cu12-bindings        10.5.0
tensorrt-cu12-libs            10.5.0
termcolor                     1.1.0
terminado                     0.17.1
terminaltables                3.1.10
thop                          0.1.1.post2209072238
threadpoolctl                 3.2.0
tifffile                      2023.4.12
tiktoken                      0.7.0
timm                          0.6.13
tinycss2                      1.2.1
tokenizers                    0.13.3
toml                          0.10.2
tomli                         2.0.1
toolz                         0.12.0
torch                         2.1.0+cu118
torchaudio                    2.1.0+cu118
torchextractor                0.3.0
torchvision                   0.16.0+cu118
tornado                       6.2
tqdm                          4.66.5
traitlets                     5.7.1
transformers                  4.26.0
transformers-stream-generator 0.0.5
triton                        2.1.0
typeshed_client               2.7.0
typing                        3.7.4.3
typing_extensions             4.12.2
tzdata                        2023.3
uc-micro-py                   1.0.3
uff                           0.6.9
urllib3                       2.2.3
uvicorn                       0.23.2
vit-pytorch                   1.4.5
watchdog                      4.0.1
wavedrom                      2.0.3.post3
wcwidth                       0.2.13
webencodings                  0.5.1
websocket-client              0.58.0
websockets                    11.0.3
Werkzeug                      3.0.1
wget                          3.2
wheel                         0.38.4
widgetsnbextension            4.0.5
xdoctest                      1.1.1
xxhash                        3.4.1
y-py                          0.5.9
yacs                          0.1.8
yapf                          0.40.0
yarl                          1.9.4
yfinance                      0.2.40
ypy-websocket                 0.8.2
zipp                          3.20.2
zstandard                     0.23.0
```

**Comment by czczup** [comment]
@Weiyun1025 看看这个问题？

**Comment by Weiyun1025** [comment]
请问你用的是我们现在开源的这套代码吗，这个问题是之前模型没有设置`supports_gradient_checkpointing = True`导致，但是[现在的代码](https://github.com/OpenGVLab/InternVL/blob/main/internvl_chat/internvl/model/internvl_chat/modeling_internvl_chat.py#L46C5-L46C43)里已经加上了这个设置

**Comment by wellhowtosay** [comment]
> 请问你用的是我们现在开源的这套代码吗，这个问题是之前模型没有设置`supports_gradient_checkpointing = True`导致，但是[现在的代码](https://github.com/OpenGVLab/InternVL/blob/main/internvl_chat/internvl/model/internvl_chat/modeling_internvl_chat.py#L46C5-L46C43)里已经加上了这个设置

可以给huggingface的代码也更新了么
